### PR TITLE
ENH: Hotelling's Two Sample Mean Test

### PR DESCRIPTION
**Pull request by w31ha0**
- [ ] closes #xxxx
- [ ] tests added / passed. 
- [ ] code/documentation is well formatted.  
- [ ] properly formatted commit message. 

Hotelling multivariate mean test for 2 sample based on following [link](http://www.real-statistics.com/multivariate-statistics/hotellings-t-square-statistic/hotellings-t-square-independent-samples/) and previously created issue #6692.

### CODE REVIEW

**Comment by w31ha0** [comment]
Not sure why i'm getting test failures when running the tests locally is passing all.

**Comment by bashtage** [comment]
Fails are unrelated to this.

File: statsmodels/stats/multivariate.py [hunk 1]
```diff
@@ -59,6 +59,52 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+        second sample data with observations in rows and variables in columns
+    return_results : bool
+        If true, then a results instance is returned. If False, then only
+        the test statistic and pvalue are returned.
+
+    Returns
+    -------
+    results : instance of a results class with attributes
+        statistic, pvalue, t2 and df
+    (statistic, pvalue) : tuple
```
Comment (bashtage) [inline statsmodels/stats/multivariate.py:79]: Return should always be the same, so please update.

File: statsmodels/stats/multivariate.py [hunk 2]
```diff
@@ -59,6 +59,52 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+        If return_results is false, then only the test statistic and the
+        pvalue are returned.
+
+    """
+    x = np.asarray(data_x)
```
Comment (bashtage) [inline statsmodels/stats/multivariate.py:84]: Use `array_like` and specif the dimension requirements.

File: statsmodels/stats/multivariate.py [hunk 3]
```diff
@@ -59,6 +59,52 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+    y = np.asarray(data_y)
```
Comment (bashtage) [inline statsmodels/stats/multivariate.py:85]: Same

File: statsmodels/stats/multivariate.py [hunk 4]
```diff
@@ -59,6 +59,52 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+    nobs_x, k_vars = x.shape
+    nobs_y, k_vars = y.shape
+    mean_x = x.mean(0)
+    mean_y = y.mean(0)
+    cov_x = np.cov(x, rowvar=False, ddof=1)
+    cov_y = np.cov(y, rowvar=False, ddof=1)
+    combined_cov = ((nobs_x - 1) * cov_x + (nobs_y - 1) * cov_y) / (nobs_x - 1 + nobs_y - 1)
+    diff = mean_x - mean_y
+    t2 = (nobs_x * nobs_y) / (nobs_x + nobs_y) * diff.dot(np.linalg.solve(combined_cov, diff))
```
Comment (bashtage) [inline statsmodels/stats/multivariate.py:94]: Use `@` rather than .dot

File: statsmodels/stats/multivariate.py [hunk 5]
```diff
@@ -59,6 +59,52 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+    factor = ((nobs_x + nobs_y - 2) * k_vars) / (nobs_x + nobs_y - k_vars - 1)
+    statistic = t2 / factor
+    df = (k_vars, nobs_x + nobs_y - 1 - k_vars)
+    pvalue = stats.f.sf(statistic, df[0], df[1])
+    if return_results:
```
Comment (bashtage) [inline statsmodels/stats/multivariate.py:99]: Always return an instance.

File: statsmodels/stats/multivariate.py (hunk 2 above)
Comment (w31ha0) [reply statsmodels/stats/multivariate.py:84]: sorry not quite sure what you mean here by array_like. Do you have an example? I thought the dimension would be something like sample size x number of variables? Which can change according to the data?

Comment (bashtage) [reply statsmodels/stats/multivariate.py:84]: https://github.com/statsmodels/statsmodels/blob/0da7220028e1e792a864df7216960ad148bdf642/statsmodels/tsa/stattools.py#L1239

**Comment by w31ha0** [comment]
> Can be made simpler and more robust with a few changes.

Done with all the comments.

**Comment by josef-pkt** [comment]
we don't use x and y for two sample tests
my convention is numbered sample x1, x2

File: statsmodels/stats/multivariate.py [hunk 6]
```diff
@@ -59,6 +59,41 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+    cov_x = np.cov(x, rowvar=False, ddof=1)
+    cov_y = np.cov(y, rowvar=False, ddof=1)
+    combined_cov = ((nobs_x - 1) * cov_x + (nobs_y - 1) * cov_y) / (nobs_x - 1 + nobs_y - 1)
+    diff = mean_x - mean_y
+    t2 = (nobs_x * nobs_y) / (nobs_x + nobs_y) * diff @ (np.linalg.solve(combined_cov, diff))
+    factor = ((nobs_x + nobs_y - 2) * k_vars) / (nobs_x + nobs_y - k_vars - 1)
+    statistic = t2 / factor
+    df = (k_vars, nobs_x + nobs_y - 1 - k_vars)
+    pvalue = stats.f.sf(statistic, df[0], df[1])
+    return HolderTuple(statistic=statistic,
```
Comment (josef-pkt) [inline statsmodels/stats/multivariate.py:92]: missing `    if return_results:` for conditional return given the keyword option

File: statsmodels/stats/multivariate.py [hunk 7]
```diff
@@ -59,6 +59,41 @@ def test_mvmean(data, mean_null=0, return_results=True):
...
+    """
+    x = array_like(x, "x", ndim=2)
+    y = array_like(y, "x", ndim=2)
+    nobs_x, k_vars = x.shape
+    nobs_y, k_vars = y.shape
+    mean_x = x.mean(0)
+    mean_y = y.mean(0)
+    cov_x = np.cov(x, rowvar=False, ddof=1)
+    cov_y = np.cov(y, rowvar=False, ddof=1)
+    combined_cov = ((nobs_x - 1) * cov_x + (nobs_y - 1) * cov_y) / (nobs_x - 1 + nobs_y - 1)
```
Comment (josef-pkt) [inline statsmodels/stats/multivariate.py:85]: you could use a temp variable `nobs_t = nobs_x + nobs_y`
I use nobs1, nobs2 in two sample case, also mean1, mean2, ...

File: statsmodels/stats/multivariate.py [hunk 8]
```diff
@@ -59,6 +59,41 @@ def test_mvmean(data, mean_null=0, return_results=True):
     else:
         return statistic, pvalue
 
+def test_mvmean_two_sample(x, y, return_results=True):
```
Comment (josef-pkt) [inline statsmodels/stats/multivariate.py:62]: my recent convention uses 2indep as postfix `test_mvmean_2indep`
(used in rates and proportions)

**Comment by josef-pkt** [comment]
Thanks for the contribution. 
looks good overall

my comments are mainly to trying to get to a more consistent naming convention across hypothesis tests and inference function. 

extension possible for future PR which would be backwards compatible with this. 
I don't remember details right now or how it applies in multivariate case
- add a `value` argument for Null `mean1 - mean2 - value = 0`
- add `use_var` option for when we don't want to assume identical cov (similar to t-test and Welch Anova)

File: statsmodels/stats/multivariate.py (hunk 6 above)
Comment (w31ha0) [reply statsmodels/stats/multivariate.py:92]: think @bashtage wanted to remove the keyword option and always return the full result set.

Comment (josef-pkt) [reply statsmodels/stats/multivariate.py:92]: that's fine also, either way, I'm still ambiguous about this as long as results instance is default.

I still prefer having the no-results option for functions where I think I might use bootstrap, simulations or compute it over a large set of cases.

**Comment by w31ha0** [comment]
> Thanks for the contribution.
> looks good overall
> 
> my comments are mainly to trying to get to a more consistent naming convention across hypothesis tests and inference function.
> 
> extension possible for future PR which would be backwards compatible with this.
> I don't remember details right now or how it applies in multivariate case
> 
> * add a `value` argument for Null `mean1 - mean2 - value = 0`
> * add `use_var` option for when we don't want to assume identical cov (similar to t-test and Welch Anova)

Done with the comments except for the no-results option.

**Comment by bashtage** [comment]
Thanks you.

**Pull request merged**